File: frontend/lib/utils/logger.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
/**
 * Production-ready logging utility with different log levels and structured output
 */

export enum LogLevel {
  DEBUG = 0,
  INFO = 1,
  WARN = 2,
  ERROR = 3,
}

export interface LogEntry {
  timestamp: string;
  level: LogLevel;
  message: string;
  context?: Record<string, any>;
  error?: Error;
  userId?: string;
  sessionId?: string;
  component?: string;
}

class Logger {
  private logLevel: LogLevel;
  private sessionId: string;
  private userId?: string;

  constructor() {
    // Set log level based on environment
    this.logLevel = process.env.NODE_ENV === 'production' ? LogLevel.WARN : LogLevel.DEBUG;
    this.sessionId = this.generateSessionId();
  }

  private generateSessionId(): string {
    return `session_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  setUserId(userId: string) {
    this.userId = userId;
  }

  private shouldLog(level: LogLevel): boolean {
    return level >= this.logLevel;
  }

  private formatLogEntry(entry: LogEntry): string {
    const { timestamp, level, message, context, error, component } = entry;
    const levelName = LogLevel[level];
    
    let logString = `[${timestamp}] ${levelName}`;
    
    if (component) {
      logString += ` [${component}]`;
    }
    
    logString += `: ${message}`;
    
    if (context && Object.keys(context).length > 0) {
      logString += ` | Context: ${JSON.stringify(context)}`;
    }
    
    if (error) {
      logString += ` | Error: ${error.message}`;
      if (error.stack && process.env.NODE_ENV !== 'production') {
        logString += `\nStack: ${error.stack}`;
      }
    }
    
    return logString;
  }

  private createLogEntry(
    level: LogLevel,
    message: string,
    context?: Record<string, any>,
    error?: Error,
    component?: string
  ): LogEntry {
    return {
      timestamp: new Date().toISOString(),
      level,
      message,
      context,
      error,
      userId: this.userId,
      sessionId: this.sessionId,
      component,
    };
  }

  private writeLog(entry: LogEntry) {
    if (!this.shouldLog(entry.level)) return;

    const formattedLog = this.formatLogEntry(entry);
    
    // Console output with appropriate method
    switch (entry.level) {
      case LogLevel.DEBUG:
        console.debug(formattedLog);
        break;
      case LogLevel.INFO:
        console.info(formattedLog);
        break;
      case LogLevel.WARN:
        console.warn(formattedLog);
        break;
      case LogLevel.ERROR:
        console.error(formattedLog);
        break;
    }

    // In production, send to monitoring service
    if (process.env.NODE_ENV === 'production' && entry.level >= LogLevel.ERROR) {
      this.sendToMonitoringService(entry);
    }

    // Store in local storage for debugging (development only)
    if (process.env.NODE_ENV === 'development') {
      this.storeLogLocally(entry);
    }
  }

  private sendToMonitoringService(entry: LogEntry) {
    // TODO: Integrate with monitoring service (e.g., Sentry, LogRocket, etc.)
    // This is a placeholder for production monitoring integration
    try {
      // Example: Sentry.captureException(entry.error, { extra: entry.context });
      console.warn('Production monitoring not configured. Log entry:', entry);
    } catch (error) {
      console.error('Failed to send log to monitoring service:', error instanceof Error ? error.message : String(error));
    }
  }

  private storeLogLocally(entry: LogEntry) {
    try {
      if (typeof window !== 'undefined' && window.localStorage) {
        const logs = JSON.parse(localStorage.getItem('polytrader_logs') || '[]');
        logs.push(entry);
        
        // Keep only last 100 logs to prevent storage overflow
        if (logs.length > 100) {
          logs.splice(0, logs.length - 100);
        }
        
        localStorage.setItem('polytrader_logs', JSON.stringify(logs));
      }
    } catch (error) {
      console.warn('Failed to store log locally:', error);
    }
  }

  debug(message: string, context?: Record<string, any>, component?: string) {
    const entry = this.createLogEntry(LogLevel.DEBUG, message, context, undefined, component);
    this.writeLog(entry);
  }

  info(message: string, context?: Record<string, any>, component?: string) {
    const entry = this.createLogEntry(LogLevel.INFO, message, context, undefined, component);
    this.writeLog(entry);
  }

  warn(message: string, context?: Record<string, any>, component?: string) {
    const entry = this.createLogEntry(LogLevel.WARN, message, context, undefined, component);
    this.writeLog(entry);
  }

  error(message: string, error?: Error, context?: Record<string, any>, component?: string) {
    const entry = this.createLogEntry(LogLevel.ERROR, message, context, error, component);
    this.writeLog(entry);
  }

  // Agent-specific logging methods
  agentStart(agentType: string, marketId: string, context?: Record<string, any>) {
    this.info(`Agent ${agentType} started`, { 
      marketId, 
      agentType,
      ...context 
    }, 'AgentSystem');
  }

  agentComplete(agentType: string, marketId: string, duration: number, context?: Record<string, any>) {
    this.info(`Agent ${agentType} completed`, { 
      marketId, 
      agentType, 
      duration,
      ...context 
    }, 'AgentSystem');
  }

  agentError(agentType: string, marketId: string, error: Error, context?: Record<string, any>) {
    this.error(`Agent ${agentType} failed`, error, { 
      marketId, 
      agentType,
      ...context 
    }, 'AgentSystem');
  }

  // Stream-specific logging methods
  streamStart(marketId: string, streamType: 'real' | 'mock') {
    this.info('Stream started', { marketId, streamType }, 'StreamSystem');
  }

  streamEnd(marketId: string, streamType: 'real' | 'mock', duration: number, eventCount: number) {
    this.info('Stream ended', { 
      marketId, 
      streamType, 
      duration, 
      eventCount 
    }, 'StreamSystem');
  }

  streamError(marketId: string, error: Error, context?: Record<string, any>) {
    this.error('Stream error', error, { marketId, ...context }, 'StreamSystem');
  }

  // Trade-specific logging methods
  tradeDecision(marketId: string, decision: any, context?: Record<string, any>) {
    this.info('Trade decision made', { 
      marketId, 
      side: decision.side,
      outcome: decision.outcome,
      size: decision.size,
      confidence: decision.confidence,
      ...context 
    }, 'TradeSystem');
  }

  tradeExecution(marketId: string, orderData: any, context?: Record<string, any>) {
    this.info('Trade executed', { 
      marketId, 
      orderId: orderData.orderID,
      status: orderData.status,
      ...context 
    }, 'TradeSystem');
  }

  tradeError(marketId: string, error: Error, context?: Record<string, any>) {
    this.error('Trade error', error, { marketId, ...context }, 'TradeSystem');
  }

  // Data validation logging
  validationError(component: string, field: string, received: any, expected: string) {
    this.warn('Data validation failed', { 
      field, 
      received: typeof received === 'object' ? JSON.stringify(received) : received,
      expected 
    }, component);
  }

  // Performance monitoring
  performanceMetric(operation: string, duration: number, context?: Record<string, any>) {
    this.info(`Performance: ${operation}`, { 
      duration, 
      operation,
      ...context 
    }, 'Performance');
  }

  // Get logs for debugging (development only)
  getLogs(): LogEntry[] {
    if (typeof window !== 'undefined' && window.localStorage) {
      try {
        return JSON.parse(localStorage.getItem('polytrader_logs') || '[]');
      } catch (error) {
        this.warn('Failed to retrieve logs from localStorage', { error: error instanceof Error ? error.message : String(error) });
        return [];
      }
    }
    return [];
  }

  // Clear logs (development only)
  clearLogs() {
    if (typeof window !== 'undefined' && window.localStorage) {
      localStorage.removeItem('polytrader_logs');
      this.info('Logs cleared');
    }
  }
}

// Create singleton instance
const logger = new Logger();

export default logger;

// Export convenience functions
export const log = {
  debug: (message: string, context?: Record<string, any>, component?: string) => 
    logger.debug(message, context, component),
  info: (message: string, context?: Record<string, any>, component?: string) => 
    logger.info(message, context, component),
  warn: (message: string, context?: Record<string, any>, component?: string) => 
    logger.warn(message, context, component),
  error: (message: string, error?: Error, context?: Record<string, any>, component?: string) => 
    logger.error(message, error, context, component),
  
  // Specialized logging functions
  agent: {
    start: (agentType: string, marketId: string, context?: Record<string, any>) => 
      logger.agentStart(agentType, marketId, context),
    complete: (agentType: string, marketId: string, duration: number, context?: Record<string, any>) => 
      logger.agentComplete(agentType, marketId, duration, context),
    error: (agentType: string, marketId: string, error: Error, context?: Record<string, any>) => 
      logger.agentError(agentType, marketId, error, context),
  },
  
  stream: {
    start: (marketId: string, streamType: 'real' | 'mock') => 
      logger.streamStart(marketId, streamType),
    end: (marketId: string, streamType: 'real' | 'mock', duration: number, eventCount: number) => 
      logger.streamEnd(marketId, streamType, duration, eventCount),
    error: (marketId: string, error: Error, context?: Record<string, any>) => 
      logger.streamError(marketId, error, context),
  },
  
  trade: {
    decision: (marketId: string, decision: any, context?: Record<string, any>) => 
      logger.tradeDecision(marketId, decision, context),
    execution: (marketId: string, orderData: any, context?: Record<string, any>) => 
      logger.tradeExecution(marketId, orderData, context),
    error: (marketId: string, error: Error, context?: Record<string, any>) => 
      logger.tradeError(marketId, error, context),
  },
  
  validation: {
    error: (component: string, field: string, received: any, expected: string) => 
      logger.validationError(component, field, received, expected),
  },
  
  performance: (operation: string, duration: number, context?: Record<string, any>) => 
    logger.performanceMetric(operation, duration, context),
};

File: backend/src/polytrader/logger.py
"""
Production-ready logging utility for the PolyTrader backend
"""

import logging
import json
import time
import asyncio
from datetime import datetime
from typing import Dict, Any, Optional
from functools import wraps
import traceback
import os
import structlog

class StructuredLogger:
    """
    Structured logger for production monitoring and debugging
    """
    
    def __init__(self, name: str = "polytrader"):
        self.logger = logging.getLogger(name)
        self.setup_logging()
        
    def setup_logging(self):
        """Configure logging based on environment"""
        # Set log level based on environment
        log_level = os.getenv("LOG_LEVEL", "INFO").upper()
        self.logger.setLevel(getattr(logging, log_level, logging.INFO))
        
        # Create formatter for structured logging
        formatter = logging.Formatter(
            '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
        )
        
        # Console handler
        if not self.logger.handlers:
            console_handler = logging.StreamHandler()
            console_handler.setFormatter(formatter)
            self.logger.addHandler(console_handler)
            
            # File handler for production
            if os.getenv("NODE_ENV") == "production":
                file_handler = logging.FileHandler("polytrader.log")
                file_handler.setFormatter(formatter)
                self.logger.addHandler(file_handler)
    
    def _log_structured(self, level: str, message: str, **kwargs):
        """Log with structured data"""
        log_data = {
            "timestamp": datetime.utcnow().isoformat(),
            "message": message,
            "level": level,
            **kwargs
        }
        
        # Convert to JSON string for structured logging
        structured_message = f"{message} | {json.dumps(log_data, default=str)}"
        
        getattr(self.logger, level.lower())(structured_message)
    
    def debug(self, message: str, **kwargs):
        self._log_structured("DEBUG", message, **kwargs)
    
    def info(self, message: str, **kwargs):
        self._log_structured("INFO", message, **kwargs)
    
    def warning(self, message: str, **kwargs):
        self._log_structured("WARNING", message, **kwargs)
    
    def error(self, message: str, error: Optional[Exception] = None, **kwargs):
        if error:
            kwargs.update({
                "error_type": type(error).__name__,
                "error_message": str(error),
                "traceback": traceback.format_exc() if error else None
            })
        self._log_structured("ERROR", message, **kwargs)
    
    def agent_start(self, agent_type: str, market_id: str, **kwargs):
        """Log agent start"""
        self.info(
            f"Agent {agent_type} started",
            agent_type=agent_type,
            market_id=market_id,
            **kwargs
        )
    
    def agent_complete(self, agent_type: str, market_id: str, duration: float, **kwargs):
        """Log agent completion"""
        self.info(
            f"Agent {agent_type} completed",
            agent_type=agent_type,
            market_id=market_id,
            duration=duration,
            **kwargs
        )
    
    def agent_error(self, agent_type: str, market_id: str, error: Exception, **kwargs):
        """Log agent error"""
        self.error(
            f"Agent {agent_type} failed",
            error=error,
            agent_type=agent_type,
            market_id=market_id,
            **kwargs
        )
    
    def trade_decision(self, market_id: str, decision: Dict[str, Any], **kwargs):
        """Log trade decision"""
        self.info(
            "Trade decision made",
            market_id=market_id,
            side=decision.get("side"),
            outcome=decision.get("outcome"),
            size=decision.get("size"),
            confidence=decision.get("confidence"),
            **kwargs
        )
    
    def trade_execution(self, market_id: str, order_data: Dict[str, Any], **kwargs):
        """Log trade execution"""
        self.info(
            "Trade executed",
            market_id=market_id,
            order_id=order_data.get("orderID"),
            status=order_data.get("status"),
            **kwargs
        )
    
    def validation_error(self, component: str, field: str, received: Any, expected: str):
        """Log validation error"""
        self.warning(
            "Data validation failed",
            component=component,
            field=field,
            received=str(received)[:100],  # Truncate long values
            expected=expected
        )
    
    def performance_metric(self, operation: str, duration: float, **kwargs):
        """Log performance metrics"""
        self.info(
            f"Performance: {operation}",
            operation=operation,
            duration=duration,
            **kwargs
        )

# Create singleton logger instance
logger = StructuredLogger()

def log_agent_execution(agent_type: str):
    """Decorator to log agent execution time and errors"""
    def decorator(func):
        if asyncio.iscoroutinefunction(func):
            @wraps(func)
            async def async_wrapper(state, *args, **kwargs):
                market_id = getattr(state, 'market_id', 'unknown')
                start_time = time.time()
                
                logger.agent_start(agent_type, market_id)
                
                try:
                    result = await func(state, *args, **kwargs)
                    duration = time.time() - start_time
                    logger.agent_complete(agent_type, market_id, duration)
                    return result
                except Exception as e:
                    duration = time.time() - start_time
                    logger.agent_error(agent_type, market_id, e, duration=duration)
                    raise
            
            return async_wrapper
        else:
            @wraps(func)
            def wrapper(state, *args, **kwargs):
                market_id = getattr(state, 'market_id', 'unknown')
                start_time = time.time()
                
                logger.agent_start(agent_type, market_id)
                
                try:
                    result = func(state, *args, **kwargs)
                    duration = time.time() - start_time
                    logger.agent_complete(agent_type, market_id, duration)
                    return result
                except Exception as e:
                    duration = time.time() - start_time
                    logger.agent_error(agent_type, market_id, e, duration=duration)
                    raise
            
            return wrapper
    return decorator

def log_performance(operation: str):
    """Decorator to log performance metrics"""
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            start_time = time.time()
            
            try:
                result = func(*args, **kwargs)
                duration = time.time() - start_time
                logger.performance_metric(operation, duration)
                return result
            except Exception as e:
                duration = time.time() - start_time
                logger.performance_metric(operation, duration, error=str(e))
                raise
        
        return wrapper
    return decorator

# Export convenience functions
def log_info(message: str, **kwargs):
    logger.info(message, **kwargs)

def log_error(message: str, error: Optional[Exception] = None, **kwargs):
    logger.error(message, error=error, **kwargs)

def log_warning(message: str, **kwargs):
    logger.warning(message, **kwargs)

def log_debug(message: str, **kwargs):
    logger.debug(message, **kwargs)
